subsys/bluetooth/services/ots/ots.c

```diff
@@ -271,63 +288,90 @@ static ssize_t ots_obj_prop_read(struct bt_conn *conn,
 
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ) && ots->dir_list &&
 	    ots->dir_list->dir_list_obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
 		LOG_DBG("Directory Listing Object is being read");
 		return -EBUSY;
 	}
 
-	name_len = strlen(obj_init->name);
-
-	CHECKIF(name_len == 0 || name_len > CONFIG_BT_OTS_OBJ_MAX_NAME_LEN) {
-		LOG_DBG("Invalid name length %zu", name_len);
-		return -EINVAL;
-	}
-
-	CHECKIF(!ots_obj_validate_prop_against_oacp(obj_init->props, ots->features.oacp)) {
-		LOG_DBG("Object properties (0x%04X) are not a subset of OACP (0x%04X)",
-				obj_init->props, ots->features.oacp);
-		return -ENOTSUP;
-	}
-
-	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &obj);
+	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &new_obj);
 	if (err) {
 		LOG_ERR("No space available in the object manager");
 		return err;
 	}
 
-	/* Initialize object. */
-	memcpy(&obj->metadata, obj_init, sizeof(obj->metadata));
-
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
-		bt_ots_dir_list_obj_add(ots->dir_list, ots->obj_manager, ots->cur_obj, obj);
+		bt_ots_dir_list_obj_add(ots->dir_list, ots->obj_manager, ots->cur_obj, new_obj);
 	}
 
-	/* Request object data. */
+	(void)memset(&created_param, 0, sizeof(created_param));
+
 	if (ots->cb->obj_created) {
-		err = ots->cb->obj_created(ots, NULL, obj->id, obj_init);
+		err = ots->cb->obj_created(ots, NULL, new_obj->id, param, &created_param);
+
 		if (err) {
-			bt_gatt_ots_obj_manager_obj_delete(obj);
+			bt_gatt_ots_obj_manager_obj_delete(new_obj);
 
 			if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ)) {
 				bt_ots_dir_list_obj_remove(ots->dir_list, ots->obj_manager,
-							   ots->cur_obj, obj);
+							   ots->cur_obj, new_obj);
 			}
 
 			return err;
 		}
 	}
 
+	new_obj->metadata.type = param->type;
+
+	new_obj->metadata.name = created_param.name;
+	new_obj->metadata.size = created_param.size;
+	new_obj->metadata.props = created_param.props;
+
+	__ASSERT(ots_obj_validate_prop_against_oacp(new_obj->metadata.props, ots->features.oacp),
+		 "Object properties (0x%04X) are not a subset of OACP (0x%04X)",
+		 new_obj->metadata.props, ots->features.oacp);
+
+	__ASSERT(new_obj->metadata.name != NULL,
+		 "Object name must be set by application after object creation.");
+
+
+	if (obj) {
+		*obj = new_obj;
+	}
+
 	return 0;
 }
 
+int bt_ots_obj_add(struct bt_ots *ots, const struct bt_ots_obj_add_param *param)
+{
+	int err;
+	size_t name_len;
+	struct bt_gatt_ots_object *obj;
+
+	err = bt_ots_obj_add_internal(ots, NULL, param, &obj);
+	if (err) {
+		return err;
+	}
+
+	name_len = strlen(obj->metadata.name);
+	__ASSERT(name_len > 0 && name_len <= CONFIG_BT_OTS_OBJ_MAX_NAME_LEN,
+		"Invalid name length %zu", name_len);
+	__ASSERT(obj->metadata.size.alloc >= param->size,
+		 "Object allocated size must be greater than or equal to requested size.");
+	__ASSERT(obj->metadata.size.cur <= param->size,
+		 "Object current size must be less than or equal to requested size.");
```

Comment:
Could these be moved to bt_ots_obj_add_internal?

Reply:
The only one that can be moved is the allocated size assertion. The other fields have different behaviors depending on whether this is a client or a server created object.

I wanted to keep all the OTS standard dictated metadata initialization together in `oacp_create_proc_validate` so I thought the duplication of the allocated size assertion was worth keeping those together. I am happy to move the one assertion if you think that is better.

Reply:
You are correct :) 

I think the metadata size checks could be moved (might as well fail as early as possible, right?).